cmd/sharded-test-server/virtual.go

```diff
@@ -27,30 +27,106 @@ import (
+	}
+	servingKeyFile := filepath.Join(workDirPath, fmt.Sprintf(".kcp-virtual-workspaces-%d/apiserver.key", index))
+	servingCertFile := filepath.Join(workDirPath, fmt.Sprintf(".kcp-virtual-workspaces-%d/apiserver.crt", index))
+	if err := cert.WriteCertConfigFile(servingCertFile, servingKeyFile); err != nil {
+		return nil, fmt.Errorf("failed to write server cert: %w", err)
+	}
+
+	// create client cert used to talk to kcp
+	vwClientCert := filepath.Join(workDirPath, fmt.Sprintf(".kcp-virtual-workspaces-%d/shard-client-cert.crt", index))
+	vwClientCertKey := filepath.Join(workDirPath, fmt.Sprintf(".kcp-virtual-workspaces-%d/shard-client-cert.key", index))
+	shardUser := &user.DefaultInfo{Name: fmt.Sprintf("kcp-vw-%d", index), Groups: []string{"system:masters"}}
+	_, err = clientCA.MakeClientCertificate(vwClientCert, vwClientCertKey, shardUser, 365)
+	if err != nil {
+		fmt.Printf("failed to create vw client cert: %v\n", err)
+		os.Exit(1)
+	}
+
+	servingCAPath, err := filepath.Abs(filepath.Join(workDirPath, ".kcp/serving-ca.crt"))
+	if err != nil {
+		fmt.Printf("error getting absolute path for %q: %v\n", filepath.Join(workDirPath, ".kcp/serving-ca.crt"), err)
+		os.Exit(1)
+	}
+	vwClientCertPath, err := filepath.Abs(vwClientCert)
+	if err != nil {
+		fmt.Printf("error getting absolute path for %q: %v\n", vwClientCert, err)
+		os.Exit(1)
+	}
+	vwClientCertKeyPath, err := filepath.Abs(vwClientCertKey)
+	if err != nil {
+		fmt.Printf("error getting absolute path for %q: %v\n", vwClientCertKey, err)
+		os.Exit(1)
+	}
+
+	virtualWorkspaceKubeConfig := clientcmdapi.Config{
+		Clusters: map[string]*clientcmdapi.Cluster{
+			"shard": {
+				Server:               fmt.Sprintf("https://localhost:%d/clusters/system:admin", 6444+index),
+				CertificateAuthority: servingCAPath,
+			},
+		},
+		Contexts: map[string]*clientcmdapi.Context{
+			"shard": {
+				Cluster:  "shard",
+				AuthInfo: "virtualworkspace",
+			},
+		},
+		AuthInfos: map[string]*clientcmdapi.AuthInfo{
+			"virtualworkspace": {
+				ClientCertificate: vwClientCertPath,
+				ClientKey:         vwClientCertKeyPath,
+			},
+		},
+		CurrentContext: "shard",
+	}
+	kubeconfigPath := filepath.Join(workDirPath, fmt.Sprintf(".kcp-virtual-workspaces-%d/virtualworkspace.kubeconfig", index))
+	err = clientcmd.WriteToFile(virtualWorkspaceKubeConfig, kubeconfigPath)
+	if err != nil {
+		fmt.Printf("failed to write vw kubeconfig: %v", err)
+		os.Exit(1)
+	}
+
 	authenticationKubeconfigPath := filepath.Join(workDirPath, fmt.Sprintf(".kcp-%d", index), "admin.kubeconfig")
 	clientCAFilePath := filepath.Join(workDirPath, ".kcp", "client-ca.crt")
 
 	commandLine := framework.DirectOrGoRunCommand("virtual-workspaces")
 	commandLine = append(
 		commandLine,
-		fmt.Sprintf("--kubeconfig=%s", authenticationKubeconfigPath),
+		fmt.Sprintf("--kubeconfig=%s", filepath.Join(workDirPath, fmt.Sprintf(".kcp-virtual-workspaces-%d/virtualworkspace.kubeconfig", index))),
```

Comment:
I think we can reuse `kubeconfigPath` here?

Reply:
done